File: core/geometry_refine.py
import numpy as np


def refine_geometry(
    outer,
    label
):
    """
    Refine geometry using
    label agreement.

    For now:
    - trust label center
    - estimate residual shift
    - estimate scale agreement
    """

    outer_x, outer_y = (
        outer["center"]
    )

    label_x, label_y = (
        label["center"]
    )

    outer_radius = (
        outer["radius_px"]
    )

    label_radius = (
        label["radius_px"]
    )

    # -------------------------
    # expected label radius
    # 100mm diameter
    # -------------------------

    expected_label_radius = (
        outer_radius
        * (
            50.0
            / 152.4
        )
    )

    # -------------------------
    # center error
    # -------------------------

    dx = (
        label_x
        - outer_x
    )

    dy = (
        label_y
        - outer_y
    )

    center_error_px = float(
        np.sqrt(
            dx**2
            + dy**2
        )
    )

    # -------------------------
    # refined center
    # trust label
    # -------------------------

    refined_x = (
        label_x
    )

    refined_y = (
        label_y
    )

    # -------------------------
    # scale estimate
    # diagnostic only
    # -------------------------

    scale_factor = (
        expected_label_radius
        /
        label_radius
    )

    confidence = max(
        0.0,
        1.0
        - (
            center_error_px
            / 15.0
        )
    )

    return {
        "center_x":
            float(
                refined_x
            ),

        "center_y":
            float(
                refined_y
            ),

        "expected_label_radius_px":
            float(
                expected_label_radius
            ),

        "detected_label_radius_px":
            float(
                label_radius
            ),

        "scale_factor":
            float(
                scale_factor
            ),

        "translation_dx":
            float(dx),

        "translation_dy":
            float(dy),

        "center_error_px":
            center_error_px,

        "geometry_confidence":
            float(
                confidence
            )
    }

File: frontend/src/engine/pipeline/geometryRefine.ts
// Mirrors backend/core/geometry_refine.py → refine_geometry()
// Pure math — no OpenCV. Trusts label center over outer ellipse center.

import type { VinylGeometry, LabelResult } from '../types';

export interface RefinedGeometry {
  centerX: number;
  centerY: number;
  expectedLabelRadiusPx: number;
  detectedLabelRadiusPx: number;
  scaleFactor: number;
  translationDx: number;
  translationDy: number;
  centerErrorPx: number;
  geometryConfidence: number;
}

export function refineGeometry(outer: VinylGeometry, label: LabelResult): RefinedGeometry {
  const [outerX, outerY] = outer.center;
  const [labelX, labelY] = label.center;

  const expectedLabelRadius = outer.radiusPx * (50.0 / 152.4);
  const dx = labelX - outerX;
  const dy = labelY - outerY;
  const centerErrorPx = Math.sqrt(dx * dx + dy * dy);

  return {
    centerX:                labelX,
    centerY:                labelY,
    expectedLabelRadiusPx:  expectedLabelRadius,
    detectedLabelRadiusPx:  label.radiusPx,
    scaleFactor:            expectedLabelRadius / label.radiusPx,
    translationDx:          dx,
    translationDy:          dy,
    centerErrorPx,
    geometryConfidence:     Math.max(0.0, 1.0 - centerErrorPx / 15.0),
  };
}
